sherpa/astro/plot.py

```diff
@@ -624,7 +624,7 @@ def plot(self, overplot=False, clearwindow=True, **kwargs):
                                   **kwargs)
             count += 1
 
-        self.histo_prefs['linecolor'] = default_color
+        self.histo_prefs['color'] = default_color
```

Comment:
So, in testing this out I came across something interesting, for an odd definition of the word interesting...

With a simple script - e..g in sherpa-test-data/sherpatest

```python
from sherpa.astro.ui import *
load_data("3273.pi")
set_source(powlaw1d.pl)
plot_order(1, [1, 2])  # in this case it is not really a normal thing to do, but it is valid code
```

you get

```
ValueError: '0xaaaaea' is not a valid value for color
```

This is not ideal, but thanks to the above code not trapping for errors, it means that the `color` attribute has been set to an invalid value and so a follow-on call in the same session, like

```python
plot_order()
```

will now fail with the same error, when it would have worked if the `plot_order(1, [1, 2])` call was not made.

If you try the same script in CIAO 4.14 then you see a plot and the warning

```
WARNING: The linecolor attribute (0xaaaaea) is unused.
```

The underlying problems are not due to this PR, but the change from the `linecolor` to `color` attribute means that it is now a problem.

The easiest fix is to revert the `linecolor` to `color` change and to note this problem as an issue to be fixed.

The way to "fix" it is that I assume something is wrong in

```
            self.colors = []
            top_color = '0xffffff'
            bot_color = '0x0000bf'
            num = len(self.orders)
            jump = (int(top_color, 16) - int(bot_color, 16)) // (num + 1)
            for order in self.orders:
                self.colors.append(top_color)
                top_color = hex(int(top_color, 16) - jump)
```

which is old code (which has seen some minimal changes, presumably python-3 related) but - thanks to linecolor actually being ignored - never actually caused a problem.

Also, if this were being fixed then I'd probably suggest something like the following, just in case

```
try:
    for xlo, xhi, y, color ...
        do stuff

finally:
    self.histo_prefs['color'] = default_color
```

Reply:
Of course, this points out a failing in our tests of `plot_order`. I thought I had added a multi-response test recently, but obviously not.

Reply:
Aha, I did add a multi-order plot: `sherpa/astro/ui/tests/test_astro_ui_multi_response.py::test_plot_order_multi` but it uses `get_order_plot` and hence doesn't end up triggering the code in question (as the `plot` method doesn't get called, unlike a test that had used `plot_order`).

Reply:
The underlying problem here - and it is interesting for this PR, talking about backend agnostic issues, is that the setting of `self.colors` I show above ***assumes** that you can specify colors as `0xrrggbb` format. This worked for ChIPS but does not work for matplotlib, which uses a different syntax (I forget exactly at the moment, e.g. can you use a string with a hex value or do you have to do something "fancy").

I guess we could try and add a "translation" step here - so something like

```python
for order in self.orders:
    sef.colors.append(color_convert(top_color))
    top_color = ...
```

where `color_convert` is the convertor (I think it has to be more specific than `translate_args['color']` as we want something that given a rgb value creates the correct color value, and this is not necessarily the same as `translate_args['color']`.

Alternatively we can drop the "auto-generate" a color array, and rely on the backend to pick different colors, but there's no guarantee that the backend would do this.

Reply:
I've added a test that should check this in #1536

Reply:
I'm putting a fix into this PR by adding a new method to backends to generate a list of colors, which should work on that particular backend. Matplotlib suppots hex colors, but wit ha slighyly different syntax (`#12AF34`)